### DISCUSSION TITLE
Utilization in prefix when IP range is defined

### DISCUSSION THREAD

**Discussion by haxorof** [netbox-community/netbox#7516] [Q&A] [answered]
I was looking into using IP ranges to define where you have DHCP inside you prefixes. When you define a IP range, which becomes a child range of a prefix, then the utilization changes even though you have not created a single IP within the IP range or the prefix. Is this how it should be? I thought it would show you 0% utilization if you have no IP addresses defined within that prefix.

See example from demo site below where I just created an IP range within an existing prefix:
![image](https://user-images.githubusercontent.com/11457642/136686193-5b5447ce-e6e5-4628-9e4e-2db3f89dc6aa.png)

![image](https://user-images.githubusercontent.com/11457642/136686208-70d1f2da-bab7-45b2-b606-98982d721f17.png)

As seen above the utilization here is at 57% even though no IP addresses has been created.

**Comment by candlerb** [comment, marked as answer]
Yes, I think that's correct.  You've assigned a range of IPs to a DHCP server, which means those IPs are no longer available for any other purpose.

Of course, the number of DHCP clients which are active on your network at any time could be anywhere between 0% and 100% of that pool.  But Netbox doesn't care about that; the point is, you've *assigned* all the addresses to the pool.  You'd need to query your DHCP server(s) to determine the number of active leases, if that's what you care about. You might also want to set up monitoring, to alert when it gets close to full.  That's a reflection of the dynamic usage of your network, not the static configuration of your network.

You wouldn't normally create IP addresses within a DHCP pool in Netbox, except when you're making static assignments for specific clients from within the dynamic pool range, and pushing those static assignments out to your DHCP server config.  That shouldn't affect the overall utilization of the pool (i.e. Netbox shouldn't double-count the IP address).

**Reply by haxorof** [reply]
Thanks @candlerb!
Totally make sense when you explain it like that. In my case we do also track the DHCP addresses and DHCP reservations in NetBox so the the utilization then started me to wonder when I defined an IP range. 😊

Cheers!

**Reply by nogaff** [reply]
@candlerb Your answer assumes an IP range would only be used in the context of a DHCP pool, where the utilisation is unknown at any given moment in time and must therefore be assumed to be 100%.

But what if I want to document that a certain range is reserved for certain types of devices with static IP addresses, and only a small percentage of that range has actually been allocated so far?

Additionally, if you view the range itself, it will correctly show < 100% utilisation, while the prefix containing that range will still consider it to have 100% utilisation, so the utilisation calculation isn't even consistent between IP ranges and prefixes.

That seems like a bug to me.

However, if that was fixed then it would mean DHCP pools defined as an IP range would always show 0% utilisation, so there would also need to be a way to mark a range as a DHCP pool to force it to have 100% utilisation.

**Reply by DanSheps** [reply]
> But what if I want to document that a certain range is reserved for certain types of devices with static IP addresses, and only a small percentage of that range has actually been allocated so far?

Then you should be using individual IP's if you don't want to see the allocation and not ranges.  Or you can see the range utilization instead to gauge how much free space you have in a range.

If it bothers you, perhaps submit a FR to adjust the calculation based on the range status (Active == Calculate utilization, Reserved/Deprecated == Don't Calculate) or add a toggle such as "Is Pool" or similar.

**Reply by nogaff** [reply]
@DanSheps That's my point though. I **am** using individual IPs, but I can't **document** the ranges they fall into, without breaking the utilisation display on my prefixes, which just seems very odd to me. 

I mean, if I've got a /24 prefix and the first 100 addresses are reserved for static IP addresses, with the remainder being for DHCP, but I've only **connected** 30 static devices, then clearly my utilisation is not 100% as the prefix utilisation would imply. It should be 73%, made up of 154 DHCP addresses, plus the 30 static devices (and the .0 and .255 at the beginning and end of the prefix).

Anyway, yes, I was indeed considering submitting an FR for an "Is pool" toggle as you suggest.

**Reply by candlerb** [reply]
Can we call it something more obvious please, like "treat as fully utilized".

("Is pool" already exists in other places in Netbox, and is used for different non-obvious reasons)

**Comment by JonathonReinhart** [comment]
Perhaps this is a separate issue or discussion, but one thing that concerns me here is that IP address which are part of a range still show up in the IP address list as "available".

For example:
- Prefix: `192.168.1.0/24`
- Range: `192.168.1.51/24` - `192.168.1.225/24` (for DHCP)\


The Range *does* show up in the "Child Ranges" tab:

![image](https://user-images.githubusercontent.com/1916566/156624620-7bb288c2-4ba4-4140-b1dd-5f50d8716a3a.png)

And it *is* properly accounted for on the main "Prefix" tab:
- Range IP addresses are not counted as "available"
- Range IP addresses are accounted for in utilization

![image](https://user-images.githubusercontent.com/1916566/156624959-a5579562-d7e0-4abc-ac59-76e9ee0e0786.png)


:warning: However, the IP addresses *do* appear here as "available":

![image](https://user-images.githubusercontent.com/1916566/156624417-80d54d93-dc1b-4a97-bb2d-6532c632f8e7.png)

Nothing is preventing someone from assigning an IP address in that range, and not realizing it's reserved for DHCP pool.

**Reply by Sprinterfreak** [reply]
Yes. This is exactly the point. Addresses claimed by a range must not be listed as usable within the prefix. This f** up things already.

**Reply by candlerb** [reply]
I think there are two distinct use cases, which would need to be different API endpoints and/or GUI actions:

* Allocate the next available address within a given *prefix*. That should exclude all ranges.
* Allocate the next available address within a given *range* (e.g. "Printers" or "Switches"). Obviously, that should pick an available address from within that range.

In the GUI, I would like the prefix to show child IPs and child ranges merged into a single view.  This would solve most of the issues I think.  Available IPs within a range, and available IPs outside of any range, could be made visually distinct, e.g. by indentation and/or shading.

**Reply by JonathonReinhart** [reply]
Yes, I think you've got it @candlerb. I agree on all points.

If nothing else, the single view would give a clue as to what's going on. Right now, the ranges, being on a separate (non-primary) tab are almost invisible to a casual user.

**Reply by bluikko** [reply]
There should be an additional checkbox "Mark as used" in addition to "pool". Or perhaps the existing "pool" checkbox should cover that "DHCP" use case, isn't that what it's really for?
Ranges are used to documenting parts of IP subnets, and they must not be marked totally unavailable since we assign IP addresses inside ranges when ranges are used for documenting "purpose".

**Reply by candlerb** [reply]
> There should be an additional checkbox "Mark as used" in addition to "pool". Or perhaps the existing "pool" checkbox should cover that "DHCP" use case, isn't that what it's really for?

I don't understand what you're proposing here. "Is a pool" and "Mark utilized" are both attributes of a prefix.  Neither is an attribute of a range.

The "pool" attribute on a prefix means that the first and last IP addresses can be treated as usable, rather than being reserved for network and broadcast addresses.

**Reply by bluikko** [reply]
Oh please excuse me then I guess I read some screenshot wrong.
Perhaps such an attribute should exist for ranges then, to put this recurring debate to bed once and for all.

**Reply by MagnusHL** [reply]
Hi!

Any updates here? 
the only update I found is that you now have the status DHCP. 

I know have to create all addresses and mark as DHCP and assign by need.

**Reply by peppermintea** [reply]
Hi,

You can create a range of ip addresses that are reserved for DHCP using the Bulk Create tab.

**Reply by candlerb** [reply]
> There should be an additional checkbox "Mark as used"

In the draft [release notes](https://github.com/netbox-community/netbox/blob/feature/docs/release-notes/version-3.5.md) for forthcoming release v3.5:

* [#7947](https://github.com/netbox-community/netbox/issues/7947) - Enable marking IP ranges as fully utilized

**Reply by teixemf** [reply]
> > There should be an additional checkbox "Mark as used"
> 
> In the draft [release notes](https://github.com/netbox-community/netbox/blob/feature/docs/release-notes/version-3.5.md) for forthcoming release v3.5:
> 
> * [#7947](https://github.com/netbox-community/netbox/issues/7947) - Enable marking IP ranges as fully utilized

I tested it on the 3.5 beta demo site. It does not solve the problem raised by @JonathonReinhart 
I have a team mate who raised the very same problem today while we were discussing our strategy for migrating our current ipam solution (phpIPAM) to Netbox.

I agree on a clean solution provided by the Netbox core. One similar to what @candlerb suggested, merge the child   IPs and child ranges view.

Nevertheless, until no better solution comes up, we thought on  tackle this by using custom validators.
We need to validate the idea but we drafted it like this:
While adding a new IP Address, check if it fits any IP Range. 
If it does not, allow the IP Address creation.
If it does, check if the status of the IP Address is DHCP.
If it does not, deny the IP creation.
If it does, allow the IP creation.

**Reply by agusadjadj** [reply]
Hi ! 
Using the DHCP Ranges, i think that It is necessary to implement that the DHCP ranges are marked as used within the "IP Addresses" tab, for greater comfort when viewing the list. Maybe adding a "IPs for DHCP" in Blue (?) color , instead of "X IPs Available" (which are in green) or something like that, can fix this. Let's wait for a soon implementation of this functionality!

Kind regards

**Reply by aldemaroc** [reply]
Looks like its still an issued. Even tho IP blocks are marked as "Fully used", we still see all IPs as "Available".

**Reply by collinscochran** [reply]
Yeah, I'd love to see those appear as unavailable when looking at the prefix view.

**Reply by jronnblom** [reply]
I also voted for this suggestion. They shouldn't be visible in the Available IPs.

**Reply by Sprinterfreak** [reply]
> Yes. This is exactly the point. Addresses claimed by a range must not be listed as usable within the prefix. This f** up things already.

I personally fixed this by not using ranges at all for now. I instead create all addresses with status DHCP or Reserved. I find ranges kind of obsolet because they are totally redundant to prefixes and addresses and therefore cause lots of confusion which repeatedly lead to duplicate address assignments.

**Reply by MarcHagen** [reply]
Hooking in here #9763 is somewhat related.
Specific this comment: https://github.com/netbox-community/netbox/issues/9763#issuecomment-1245886761

It seems something in the templating need to be changed to reflect this. It needs some good thinking to change.
https://github.com/netbox-community/netbox/blob/28aee9b69a7e4b55001b3e6073c6f49b53c31afd/netbox/ipam/tables/ip.py#L50-L58

**Comment by telmich** [comment]
I was just running into the same issue, expecting a range to cover/show that an IP address is being used already, would be very natural to me.
